### ISSUE TITLE
Is QueryRequest COSE protected

### ISSUE THREAD

**Issue by dthaler** [ietf-teep/teep-protocol#251]
Section 7.2 says:

> When the ProcessTeepMessage API is invoked, the Agent first does validation as specified in [Section 4.1.2](https://www.ietf.org/archive/id/draft-ietf-teep-protocol-10.html#validation), and drops the message if it is not valid. Otherwise, processing continues as follows based on the type of message.

Thus 4.1.2 must be done independent of the type of message.  Section 4.1.2 says:

> When TEEP message is received (see the ProcessTeepMessage conceptual API defined in [[I-D.ietf-teep-architecture](https://www.ietf.org/archive/id/draft-ietf-teep-protocol-10.html#I-D.ietf-teep-architecture)] section 6.2.1), the following validation steps are performed. If any of the listed steps fail, then the TEEP message MUST be rejected.
> 
> 1. Verify that the received message is a valid CBOR object.
> 2. Verify that the message contains a COSE_Sign1 structure.
> 3. ...

Thus all messages must be wrapped in COSE_Sign1 including the QueryRequest, according to the above.

However, section 8 on cipher suites says:

> The TEEP protocol uses COSE for protection of TEEP messages. **After a QueryResponse is received**, the selected cipher suite is used by both the TAM and the TEEP Agent in all subsequent TEEP messages. That is, a cipher suite is used in both directions.

So if the selected cipher suite (which is communicated in the QueryResponse) is used only AFTER a QueryResponse is received, then what cipher suite is used to protect the QueryRequest and QueryResponse?   The TAM must support both so could simply try all supported ones on the QueryResponse.  However, the Agent need only pick one, and the TAM will not necessarily know which one the Agent supports until negotiation completes, so how can it protect the QueryRequest?

**Comment by dthaler** [comment]
Some options would be to:
A) don’t protect the QueryRequest.  This has the risk of additional security concerns.
B) make the TAM send a QueryRequest using each cipher suite until one elicits a QueryResponse.  This one doesn't work well with the transport which requires one QueryRequest in response to an HTTP request to the TAM.
C) use the Accept header in the Agent's connect request to specify the ciphersuite.  This makes the implementation up to the transport and hence can vary, increasing the complexity of writing a transport.  However this could potentially work.
D) use a separate TAM URI per ciphersuite.

**Comment by mcd500** [comment]
@dthaler 
I will review it next week, just being busy.

**Comment by ko-isobe** [comment]
New option: Multiple algorithms in one QueryRequest message

TAM can attach multiple signatures with COSE_Sign using different algorithm and key pair.

**Comment by dthaler** [comment]
> New option: Multiple algorithms in one QueryRequest message
> 
> TAM can attach multiple signatures with COSE_Sign using different algorithm and key pair.

That's a great point @ko-isobe, I actually like that better than the other options.  I will file a separate PR in the hopes we can merge that one and archive this one, and then verify at IETF which approach has consensus.

**Comment by dthaler** [comment]
Fixed in draft-11
